soroban-sdk/src/env.rs

```diff
@@ -549,6 +621,79 @@ impl Env {
         contract_id
     }
 
+    /// Checks if a top-level `require_auth` call has happened during the last
+    /// contract invocation for the given `address` and contract invocation.
+    ///
+    /// This also removes the record about the authorization, so calling this
+    /// for the second time with the same arguments will usually return `false`
+    /// (unless the identical authorization happened multiple times).
+    ///
+    /// 'Top-level call' here means that this is the first call of
+    /// `require_auth` for a given address in the call stack; it doesn't have
+    /// to coincide with the actual top-level contract invocation. For example,
+    /// if contract A doesn't use `require_auth` and then it calls contract B
+    /// that uses `require_auth`, then `verify_top_authorization` will return
+    /// `true` when verifying the contract B's `require_auth` call, but it will
+    /// return `false` if contract A makes a `require_auth` call.
+    ///
+    /// ### Examples
+    /// ```
+    /// use soroban_sdk::{contractimpl, Address, Env, IntoVal};
+    ///
+    /// pub struct Contract;
+    ///
+    /// #[contractimpl]
+    /// impl Contract {
+    ///     pub fn transfer(env: Env, address: Address, amount: i128) {
+    ///         address.require_auth((amount / 2,).into_val(&env));
+    ///     }
+    /// }
+    ///
+    /// #[test]
+    /// fn test() {
+    /// # }
+    /// # #[cfg(feature = "testutils")]
+    /// # fn main() {
+    ///     let env = Env::default();
+    ///     let contract_id = env.register_contract(None, Contract);
+    ///     let client = ContractClient::new(&env, &contract_id);
+    ///     let address = Address::random(&env);
+    ///     client.transfer(&address, &1000_i128);
+    ///     // `transfer` requires auth for (amount / 2) == (1000 / 2) == 500.
+    ///     assert!(env.verify_top_authorization(
```

Comment:
This type of testing reminds me of tests that utilize mocks, where we're trying to track if a call to the mock has occurred or not. Here we're trying to track if a call to require_authorization has occurred.

I think this testing API would be more intuitive to users if we presented it as test double or a mock or a spy, since that's effectively what it is.

But, I think we should try to find another way to make this assertion, because the error this assertion displays will be difficult to debug because it is just a yes or a no.

Could we add a new function that gets a Vec<> of require authorizations that have been called? The way this would be written in a test would then be something like this:

```rust
client.transfer(&address, &1000_i128);
assert_eq!(
    env.auth().all(),
    std::vec![
        (&address, &contract_id, "transfer", (500_i128,).into_val(&env)),
    ],
);
```

Writing the test this way has several advantages:
- If someone doesn't know what is going on they can print out all the authorizations.
- The assert_eq will print out the full list of authorizations if they don't match what is expected.

As a bonus the API of the SDK will now only present facts and data, and is less prescriptive, and causes less side-effects.

Reply:
I have mixed feelings about setting up mocks; while pretty nice in concept they're frequently tricky to setup correctly and require pretty significant library support.

I like your suggestion. The current design stems from the previous iteration where you needed to pass the whole call stack and I want to allow users to only verify a few calls they could be interested in. But If we do this for the top level, that's probably simple enough to use.

Since this requires a host change, I'll open an issue for that for now: https://github.com/stellar/rs-soroban-env/issues/663